### PR TITLE
OEL-2283: Added overridable icon_path.

### PR DESCRIPTION
**Pull request by tibi2303**
Jira issue(s):
- [D8AGE-???](https://webgate.ec.europa.eu/CITnet/jira/browse/D8AGE-???)

### CODE REVIEW

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    $variables = [];
+
+    // Call the preprocess function manually.
+    oe_bootstrap_theme_preprocess($variables);
+
+    // Extract the path set in the preprocess function.
+    $actualIconPath = $variables['bcl_icon_path'];
+
+    // Check if the paths match.
+    $this->assertEquals($expectedIconPath, $actualIconPath);
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: Not sure whats the expected value here, we have one test with one result.
There is a coalesce operator on line 58, let's make it static value to know whats the expected value, remove coalesce.
1. Assert the theme bcl_icon_path has a value like 'assets/icons/bcl-default-icons-test.svg'
2. Assert that the value is the expected value.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    $themeHandler = \Drupal::service('theme_handler');
+
+    // Switch to the base theme.
+    $this->config('system.theme')->set('default', 'oe_bootstrap_theme')->save();
+    // Retrieve the active theme.
+    $activeTheme = $this->themeManager->getActiveTheme();
+
+    // Access the bcl_icon_path from the theme info file.
+    $themeInfo = $activeTheme->getExtension()->info;
+    $bclIconPath = $themeInfo['openeuropa']['bootstrap_component_library']['bcl_icon_path'] ?? 'assets/icons/bcl-default-icons.svg';
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: Same here, we don't know if the value for bcl_icon_path is null or not.
1. Assert value of setting, if is not set check that.
2. Assert expected output, fallback.

For me fallback should be tested from child theme as is done in override, but no value is set.

You can do the same as for components_location, and use a oe_bootstrap_theme_test_subtheme2 without value.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php:1]: If the variable is managed by theme I think the test should be placed in theme tests, not module. I would add separate tests theme too, with minimum info.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    // Check if the paths match.
+    $this->assertEquals($expectedIconPath, $actualIconPath);
+  }
+
+  /**
+   * Tests the fallback when the bcl_icon_path is not set in the subtheme.
+   */
+  public function testBclIconPathFallback(): void {
+    // Use the ThemeHandler service to retrieve the base theme.
+    $themeHandler = \Drupal::service('theme_handler');
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: This variable is not used.

File: oe_bootstrap_theme.theme [hunk 4]
```diff
@@ -61,7 +61,15 @@ function oe_bootstrap_theme_theme_suggestions_views_view_table_alter(array &$sug
  * Implements hook_preprocess().
  */
 function oe_bootstrap_theme_preprocess(&$variables) {
-  $variables['bcl_icon_path'] = base_path() . \Drupal::service('extension.list.theme')->getPath('oe_bootstrap_theme') . '/assets/icons/bcl-default-icons.svg';
+  // Get the active theme and its path.
+  $theme = \Drupal::theme()->getActiveTheme();
+  $theme_path = \Drupal::service('extension.list.theme')->getPath($theme->getName());
```
Comment (AaronGilMartinez) [inline oe_bootstrap_theme.theme]:
```suggestion
  $theme_path = $theme->getPath();
```

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ */
+class BclIconPathOverrideTest extends KernelTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'system',
+    'theme_test',
+    'oe_bootstrap_theme_helper',
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: Not needed helper module, the theme does the preprocess.
Module theme_test can be removed too.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    // Initialize the theme manager service.
+    $this->themeManager = $this->container->get('theme.manager');
+
+    // Install the theme. Subsequently, this installs also the base themes.
+    $this->container->get('theme_installer')->install(['oe_bootstrap_theme', 'oe_bootstrap_theme_test_subtheme1']);
+    $this->config('system.theme')->set('default', 'oe_bootstrap_theme_test_subtheme1')->save();
+
+    // Manually include the .theme file of the base theme.
+    $themePath = \Drupal::service('extension.list.theme')->getPath('oe_bootstrap_theme');
+    require_once $themePath . '/' . 'oe_bootstrap_theme.theme';
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: Not needed.

File: oe_bootstrap_theme.theme (hunk 4 above)
Comment (AaronGilMartinez) [reply oe_bootstrap_theme.theme]: Apply everywhere, please.

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php (hunk 1 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: done.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php (hunk 2 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: done.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php:1]: done.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php (hunk 3 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: done.

File: oe_bootstrap_theme.theme (hunk 4 above)
Comment (tibi2303) [reply oe_bootstrap_theme.theme]: done.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php (hunk 5 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: done.

File: modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php (hunk 6 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/src/Kernel/BclIconPathOverrideTest.php]: done.

File: modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme1/oe_bootstrap_theme_test_subtheme1.info.yml [hunk 7]
```diff
@@ -6,3 +6,4 @@ core_version_requirement: ^10
 openeuropa:
   bootstrap_component_library:
     components_location: components/templates
+    bcl_icon_path: 'assets/icons/bcl-default-icons-test.svg'
```
Comment (AaronGilMartinez) [inline modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme1/oe_bootstrap_theme_test_subtheme1.info.yml]: We don't need this here if we create test themes.

File: oe_bootstrap_theme.theme [hunk 8]
```diff
@@ -61,7 +61,15 @@ function oe_bootstrap_theme_theme_suggestions_views_view_table_alter(array &$sug
  * Implements hook_preprocess().
  */
 function oe_bootstrap_theme_preprocess(&$variables) {
-  $variables['bcl_icon_path'] = base_path() . \Drupal::service('extension.list.theme')->getPath('oe_bootstrap_theme') . '/assets/icons/bcl-default-icons.svg';
+  // Get the active theme and its path.
+  $theme = \Drupal::theme()->getActiveTheme();
+  $theme_path = $theme->getPath($theme->getName());
```
Comment (AaronGilMartinez) [inline oe_bootstrap_theme.theme]: This method doesn't accept parameters, the test are failing which is good because is telling us something is not ok.

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Retrieve the active theme.
+    $theme = $this->themeManager->getActiveTheme();
+
+    // Access the bcl_icon_path from the theme info file.
+    $themeInfo = $theme->getExtension()->info;
+
+    // Optionally, check if the variable is empty if it exists.
+    $bclIconPath = $themeInfo['openeuropa']['bootstrap_component_library']['bcl_icon_path'] ?? NULL;
+    $this->assertEmpty($bclIconPath);
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/BclIconPathOverrideTest.php]:
```suggestion
    $this->assertNull($bclIconPath);
```

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

File: modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme1/oe_bootstrap_theme_test_subtheme1.info.yml (hunk 7 above)
Comment (tibi2303) [reply modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme1/oe_bootstrap_theme_test_subtheme1.info.yml]: done.

File: oe_bootstrap_theme.theme (hunk 8 above)
Comment (tibi2303) [reply oe_bootstrap_theme.theme]: done.

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 9 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: done.

**Review by AaronGilMartinez — DISMISSED** [review]

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  protected static $modules = [
+    'system',
+  ];
+
+  /**
+   * The theme manager service.
+   *
+   * @var \Drupal\Core\Theme\ThemeManagerInterface
+   */
+  protected ThemeManagerInterface $themeManager;
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: We don't set services as properties, unless very specific scenarios.

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+  /**
+   * Set up the test environment.
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    // Initialize the theme manager service.
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: We are missing the file_scan_ignore_directories behaviour as per \Drupal\Tests\oe_bootstrap_theme_helper\Kernel\BclComponentsOverrideTest::setUp().

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    $this->themeManager = $this->container->get('theme.manager');
+
+    // Install the theme. Subsequently, this installs also the base themes.
+    $this->container->get('theme_installer')->install(['oe_bootstrap_theme', 'oe_bootstrap_theme_test_subtheme']);
+    $this->config('system.theme')->set('default', 'oe_bootstrap_theme_test_subtheme')->save();
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: There's no need to do this here.

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+  /**
+   * Tests that the bcl_icon_path is correctly overridden in the subtheme.
+   */
+  public function testBclIconPathOverride(): void {
+    // Retrieve the active theme.
+    $theme = $this->themeManager->getActiveTheme();
+
+    // Access the bcl_icon_path from the theme info file.
+    $themeInfo = $theme->getExtension()->info;
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: snake_case, no CamelCase.

File: tests/theme/oe_bootstrap_theme_test_subtheme/oe_bootstrap_theme_test_subtheme.info.yml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+name: OpenEuropa Bootstrap Theme Test Subtheme
```
Comment (brummbar) [inline tests/theme/oe_bootstrap_theme_test_subtheme/oe_bootstrap_theme_test_subtheme.info.yml]: Didn't we say about using the already existing test themes?

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Retrieve the active theme.
+    $theme = $this->themeManager->getActiveTheme();
+
+    // Access the bcl_icon_path from the theme info file.
+    $themeInfo = $theme->getExtension()->info;
+
+    // Optionally, check if the variable is empty if it exists.
+    $bclIconPath = $themeInfo['openeuropa']['bootstrap_component_library']['bcl_icon_path'] ?? NULL;
+    $this->assertNull($bclIconPath);
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: We don't need to run these checks, we don't need to test our test data. We only need to run the preprocess and compare it with a static value that we know it's in the theme info file, like \Drupal\Tests\oe_bootstrap_theme_helper\Kernel\BclComponentsOverrideTest::bclComponentsOverridingTestCasesProvider.

**Review by brummbar — CHANGES_REQUESTED** [review]
Let's use a dataprovider with 3 scenarios like \Drupal\Tests\oe_bootstrap_theme_helper\Kernel\BclComponentsOverrideTest::bclComponentsOverridingTestCasesProvider.

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 10 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: removed it

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 11 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: added

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 12 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: yep, refactored the tests with data provider

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 13 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: redone it

File: tests/theme/oe_bootstrap_theme_test_subtheme/oe_bootstrap_theme_test_subtheme.info.yml (hunk 14 above)
Comment (tibi2303) [reply tests/theme/oe_bootstrap_theme_test_subtheme/oe_bootstrap_theme_test_subtheme.info.yml]: yep

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 15 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: refactored it

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    ];
+    new Settings($settings);
+  }
+
+  /**
+   * Tests that the bcl_icon_path is correctly handled in the theme.
+   *
+   * @param string $theme
+   *   The theme to set as default.
+   * @param string|null $expected_icon_path
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/BclIconPathOverrideTest.php]:
```suggestion
   * @param string $expected_icon_path
```

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+   * @dataProvider bclIconPathTestCasesProvider
+   */
+  public function testBclIconPath(string $theme, string $expected_icon_path): void {
+    // Install the theme. Subsequently, this installs also the base themes.
+    $this->container->get('theme_installer')->install([$theme]);
+    $this->config('system.theme')->set('default', $theme)->save();
+
+    // Retrieve the active theme.
+    $themeManager = $this->container->get('theme.manager');
+    $active_theme = $themeManager->getActiveTheme();
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/BclIconPathOverrideTest.php]:
```suggestion
    \Drupal::theme()->getActiveTheme();
```

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]
Small changes

**Review by AaronGilMartinez — DISMISSED** [review]

File: tests/src/Kernel/BclIconPathOverrideTest.php [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  public function bclIconPathTestCasesProvider(): array {
+    return [
+      'subtheme theme with overridden path' => [
+        'oe_bootstrap_theme_test_subtheme1',
+        '/themes/custom/oe_bootstrap_theme/modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme1/assets/icons/bcl-default-icons-test.svg',
+      ],
+      'subtheme theme with fallback path' => [
+        'oe_bootstrap_theme_test_subtheme2',
+        '/themes/custom/oe_bootstrap_theme/modules/oe_bootstrap_theme_helper/tests/themes/oe_bootstrap_theme_test_subtheme2/assets/icons/bcl-default-icons.svg',
```
Comment (brummbar) [inline tests/src/Kernel/BclIconPathOverrideTest.php]: Now that we have these tests, it outlines that the current implementation is flawed.
When a theme doesn't declare the value, it should use the value of the parent theme, if present. All the way up the chain until hitting oe_bootstrap_theme.
We don't want to force subthemes to copy the svg in their folder if no override is needed.

**Review by brummbar — CHANGES_REQUESTED** [review]

File: tests/src/Kernel/BclIconPathOverrideTest.php (hunk 18 above)
Comment (tibi2303) [reply tests/src/Kernel/BclIconPathOverrideTest.php]: done.

**Review by brummbar — APPROVED** [review]

**Review by AaronGilMartinez — APPROVED** [review]
